packages/vite/src/node/server/middlewares/transform.ts

```diff
@@ -74,23 +74,29 @@ export function transformMiddleware(
       !req.url?.startsWith(CLIENT_PUBLIC_PATH) &&
       !req.url?.includes('vite/dist/client')
     ) {
-      // missing dep pending reload, hold request until reload happens
-      server._pendingReload.then(() =>
-        // If the refresh has not happened after timeout, Vite considers
-        // something unexpected has happened. In this case, Vite
-        // returns an empty response that will error.
-        setTimeout(() => {
-          // Don't do anything if response has already been sent
-          if (res.writableEnded) return
+      try {
+        // missing dep pending reload, hold request until reload happens
+        await Promise.race([
+          server._pendingReload,
+          // If the refresh has not happened after timeout, Vite considers
+          // something unexpected has happened. In this case, Vite
+          // returns an empty response that will error.
+          new Promise<void>((_, reject) =>
```

Comment:
**suggestion (non-blocking):** If I'm correct, I think we can leave away the generic

```suggestion
          new Promise((_, reject) =>
```